File: backend/arrlink/core/snapshot.py
"""Rebuild adapter-shaped `Item`/`MediaFile` objects from the poller's stored snapshot (`app_items` + `app_files`)."""

from __future__ import annotations

import json

from ..arr.base import Item, MediaFile
from ..state import State


def snapshot_items(db: State, app_id: int) -> list[Item]:
    """Every stored item for the app, with its stored files attached.

    Returns [] if the app has never been polled (no rows) — callers should
    fall back to a live fetch in that case.
    """
    files_by_item: dict[int, list[MediaFile]] = {}
    for fr in db.query(
        "SELECT f.* FROM app_files f JOIN app_items i ON i.id = f.item_id "
        "WHERE i.app_id = ? ORDER BY f.id",
        (app_id,),
    ):
        files_by_item.setdefault(fr["item_id"], []).append(
            MediaFile(
                rel_path=fr["rel_path"],
                abs_path=fr["abs_path"],
                size=fr["size"],
                mtime=fr["mtime"],
                inode=fr["inode"],
                id=fr["id"],
            )
        )

    items: list[Item] = []
    for r in db.query("SELECT * FROM app_items WHERE app_id = ? ORDER BY id", (app_id,)):
        items.append(
            Item(
                id=r["item_id"],
                title=r["title"],
                year=r["year"],
                tags=json.loads(r["tags_json"] or "[]"),
                path=r["path"] or "",
                files=files_by_item.get(r["id"], []),
                genres=json.loads(r["genres_json"] or "[]"),
                certification=r["certification"],
                collection=r["collection"],
                quality_profile_id=r["quality_profile_id"],
                quality_profile_name=r["quality_profile_name"],
                original_language=r["original_language"],
            )
        )
    return items

File: backend-ts/src/core/snapshot.ts
import { eq } from 'drizzle-orm'
import type { DbClient } from '../db/client.js'
import { appFiles, appItems } from '../db/schema.js'
import type { Item, MediaFile } from '../arr/types.js'

/** Rebuild adapter-shaped Item/MediaFile objects from the poller's stored snapshot.
 * Ported from core/snapshot.py. */

/** Every stored item, with files attached. [] if never polled -- callers fall back to live fetch. */
export function snapshotItems(db: DbClient, appId: number): Item[] {
  const itemRows = db.select().from(appItems).where(eq(appItems.appId, appId)).all()

  const filesByItemDbId = new Map<number, MediaFile[]>()
  for (const { app_files: f } of db
    .select({ app_files: appFiles })
    .from(appFiles)
    .innerJoin(appItems, eq(appFiles.itemId, appItems.id))
    .where(eq(appItems.appId, appId))
    .all()) {
    const list = filesByItemDbId.get(f.itemId)
    const file: MediaFile = {
      relPath: f.relPath,
      absPath: f.absPath,
      size: f.size,
      mtime: f.mtime,
      inode: f.inode,
      id: f.id,
    }
    if (list) list.push(file)
    else filesByItemDbId.set(f.itemId, [file])
  }

  return itemRows.map((r): Item => ({
    id: r.itemId,
    title: r.title,
    year: r.year,
    tags: JSON.parse(r.tagsJson || '[]') as string[],
    path: r.path || '',
    files: filesByItemDbId.get(r.id) ?? [],
    genres: JSON.parse(r.genresJson || '[]') as string[],
    certification: r.certification,
    collection: r.collection,
    qualityProfileId: r.qualityProfileId,
    qualityProfileName: r.qualityProfileName,
    originalLanguage: r.originalLanguage,
    audioLanguages: JSON.parse(r.audioLanguagesJson || '[]') as string[],
    studio: r.studio,
    network: r.network,
    seriesType: r.seriesType,
    videoCodec: JSON.parse(r.videoCodecJson || '[]') as string[],
    videoDynamicRange: JSON.parse(r.videoDynamicRangeJson || '[]') as string[],
    audioCodec: JSON.parse(r.audioCodecJson || '[]') as string[],
    audioChannels: JSON.parse(r.audioChannelsJson || '[]') as string[],
    rating: r.rating,
    popularity: r.popularity,
    runtime: r.runtimeMinutes,
    dbId: r.id,
  }))
}
